test/src/constraints/GapHeatConductanceAuxKernel.C

```diff
@@ -0,0 +1,77 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "GapHeatConductanceAuxKernel.h"
+
+registerMooseObject("MooseTestApp", GapHeatConductanceAuxKernel);
+
+InputParameters
+GapHeatConductanceAuxKernel::validParams()
+{
+  InputParameters params = ADMortarConstraint::validParams();
+  params.addParam<Real>("gap_conductance_constant",
+                        0.03,
+                        "The numerator in calculation of the heat transfer coefficient");
+  params.addParam<MaterialPropertyName>(
+      "slave_gap_conductance",
+      "gap_conductance",
+      "The material property name providing the gap conductance on the slave side");
+  params.addParam<MaterialPropertyName>(
+      "master_gap_conductance",
+      "gap_conductance",
+      "The material property name providing the gap conductance on the master side");
+  params.addRequiredCoupledVar(
+      "auxkernel_variable", "The auxiliary kernel variable that uses a lower-d for compute values");
+
+  return params;
+}
+
+GapHeatConductanceAuxKernel::GapHeatConductanceAuxKernel(const InputParameters & parameters)
+  : ADMortarConstraint(parameters),
+    _slave_gap_conductance(getADMaterialProperty<Real>("slave_gap_conductance")),
+    _master_gap_conductance(getNeighborADMaterialProperty<Real>("master_gap_conductance")),
+    _aux_variable(coupledValue("auxkernel_variable")),
+    _aux_variable_old(coupledValueOld("auxkernel_variable"))
+{
+}
+
+ADReal
+GapHeatConductanceAuxKernel::computeQpResidual(Moose::MortarType type)
+{
+  switch (type)
+  {
+    case Moose::MortarType::Slave:
+      return _lambda[_qp] * _test_slave[_i][_qp];
+
+    case Moose::MortarType::Master:
+      return -_lambda[_qp] * _test_master[_i][_qp];
+
+    case Moose::MortarType::Lower:
+    {
+      ADReal heat_transfer_coeff(0);
+      if (_has_master)
+      {
+        auto gap = (_phys_points_slave[_qp] - _phys_points_master[_qp]).norm();
+        heat_transfer_coeff =
+            (0.5 * (_slave_gap_conductance[_qp] + _master_gap_conductance[_qp])) / gap;
```

Comment:
`gap` will be always nonzero?

Reply:
As the tests are set up `gap` is always nonzero. Added an `assert` in case another application or test uses these ad hoc classes.